File: src/main/java/pl/training/workshop/m5/s15_compatibility/Column.java
package pl.training.workshop.m5.s15_compatibility;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/** Stabilny kontrakt sceny: adnotacja kolumny eksportu (czytana refleksją, jak w ORM/JSON/CSV). */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Column {
    String value();
}

File: typescript/src/workshop/m5/s15_compatibility/Column.ts
/**
 * Stabilny kontrakt sceny: metadane kolumny eksportu przypięte do metody - odpowiednik adnotacji
 * @Column("cena") czytanej refleksją (jak w ORM/JSON/CSV). Dekoratory TC39 nie działają jeszcze
 * w całym łańcuchu narzędzi (Node 22, vitest), więc klasa oznacza metodę w bloku statycznym:
 *
 *   static {
 *     column(this.prototype.price, 'cena');
 *   }
 *
 * Metadane są przypięte do obiektu funkcji, a więc do klasy, która metodę DEKLARUJE.
 */
const columns = new WeakMap<object, string>();

export function column(method: object, name: string): void {
  columns.set(method, name);
}

/** Nazwa kolumny metody albo undefined (odpowiednik method.getAnnotation(Column.class)). */
export function columnOf(method: unknown): string | undefined {
  return typeof method === 'function' ? columns.get(method) : undefined;
}
